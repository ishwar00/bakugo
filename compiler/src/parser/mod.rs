use miette::NamedSource;
use pest::{
    error::Error,
    iterators::{Pair, Pairs},
    Parser,
};

use crate::ast::{BakugoParsingErrorDisplay, Node, SourceFile};

#[derive(pest_derive::Parser)]
#[grammar = "bakugo.pest"]
pub struct BakugoParser;

pub fn parse_string(s: &str) -> Result<Pairs<'_, Rule>, Box<Error<Rule>>> {
    match BakugoParser::parse(Rule::SourceFile, s) {
        Ok(ast) => Ok(ast),
        Err(err) => Err(Box::new(err)),
    }
}

pub fn construct_ast(
    source: NamedSource,
    pair: Pair<'_, Rule>,
) -> Result<SourceFile<'_>, BakugoParsingErrorDisplay> {
    match SourceFile::parse(pair) {
        Ok(s) => Ok(s),
        Err(bakugo_err) => Err(BakugoParsingErrorDisplay::from_error(bakugo_err, source)),
    }
}
