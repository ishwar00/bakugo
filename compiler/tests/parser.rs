use std::fs;

use bakugo::parser::parse_string;
use insta::{assert_snapshot, assert_yaml_snapshot, glob};

#[test]
fn test_parser() {
    glob!("examples/*.bakugo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let parsed = parse_string(&input);
        match parsed {
            Ok(parsed) => assert_yaml_snapshot!(parsed),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    });
}
