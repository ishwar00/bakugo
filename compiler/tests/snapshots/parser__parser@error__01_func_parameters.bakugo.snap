---
source: tests/parser.rs
expression: parsed
input_file: tests/examples/error/01_func_parameters.bakugo
---
pos:
  - 0
  - 20
pairs:
  - pos:
      - 0
      - 20
    rule: SourceFile
    inner:
      pos:
        - 0
        - 20
      pairs:
        - pos:
            - 0
            - 18
          rule: TopLevelDecl
          inner:
            pos:
              - 0
              - 18
            pairs:
              - pos:
                  - 0
                  - 18
                rule: FunctionDecl
                inner:
                  pos:
                    - 5
                    - 17
                  pairs:
                    - pos:
                        - 5
                        - 9
                      rule: Ident
                      inner: expr
                    - pos:
                        - 9
                        - 15
                      rule: Signature
                      inner:
                        pos:
                          - 9
                          - 14
                        pairs:
                          - pos:
                              - 9
                              - 14
                            rule: Parameters
                            inner:
                              pos:
                                - 10
                                - 13
                              pairs:
                                - pos:
                                    - 10
                                    - 13
                                  rule: ParameterList
                                  inner:
                                    pos:
                                      - 10
                                      - 13
                                    pairs:
                                      - pos:
                                          - 10
                                          - 13
                                        rule: ParameterDecl
                                        inner:
                                          pos:
                                            - 10
                                            - 13
                                          pairs:
                                            - pos:
                                                - 10
                                                - 13
                                              rule: Type
                                              inner:
                                                pos:
                                                  - 10
                                                  - 13
                                                pairs:
                                                  - pos:
                                                      - 10
                                                      - 13
                                                    rule: Ident
                                                    inner: int
                    - pos:
                        - 17
                        - 17
                      rule: StatementList
                      inner: ""
        - pos:
            - 18
            - 19
          rule: Semicolon
          inner: ;
        - pos:
            - 20
            - 20
          rule: EOI
          inner: ""

